python/cudf/cudf/core/dataframe.py

```diff
@@ -283,12 +289,22 @@ def _init_from_dict_like(self, data, index=None, columns=None):
             # not in `columns`
             keys = [key for key in data.keys() if key in columns]
             data = {key: data[key] for key in keys}
+            extra_cols = [col for col in columns if col not in data.keys()]
             if keys:
                 # if keys is non-empty,
                 # add null columns for all values
                 # in `columns` that don't exist in `keys`:
-                extra_cols = [col for col in columns if col not in data.keys()]
                 data.update({key: None for key in extra_cols})
+            else:
+                # if keys is empty,
+                # it means that none of the actual keys in `data`
+                # matches with `columns`.
+                # Hence only assign `data` with `columns` as keys
+                # and their values as empty columns.
+                data = dict.fromkeys(
+                    extra_cols,
+                    cudf.core.column.column_empty(row_count=0, dtype=None),
+                )
```

Comment:
Along the same lines as Keith's previous comment, we probably also want to use a dict comprehension here. My mistake - I keep forgetting that columns are mutable:

```python
data = {key:cudf.core.column.column_empty(row_count=0, dtype=None) for key in extra_cols}
```

Reply:
Thanks Ashwin!  Changed it 👍